File: hash_table/0128_longest_consecutive_sequence.js
/*
Given an unsorted array of integers nums, return the length of the longest 
consecutive elements sequence.

You must write an algorithm that runs in O(n) time.
*/
// https://leetcode.com/problems/longest-consecutive-sequence/

// hash table + dynamic programming
// time complexity O(n)
// space complexity O(n)
var longestConsecutive = function (nums) {
  let numsSet = new Set(nums);
  let ans = 0;
  for (let num of numsSet) {
    // filter out the start of streak, optimizing performance
    if (!numsSet.has(num - 1)) {
      let currNum = num;
      let currCount = 1;
      // start counting
      // ATTN: all the numbers in while loop will be filtered out the for loop above
      while (numsSet.has(currNum + 1)) {
        currNum++;
        currCount++;
      }
      ans = Math.max(ans, currCount);
    }
  }
  return ans;
};

// make an new array as a hash table - overflow error
// time complexity O(n)
// space complexity O(N)
var longestConsecutive = function (nums) {
  if (nums.length <= 1) return nums.length;
  let max = Math.max(...nums);
  let min = Math.min(...nums);
  max = Math.max(max, Math.abs(min));
  // in JS the maximum length of an array is 2 ** 32 - 1 < 10 ** 9, will result in error
  let array = new Array(max * 2).fill(0);
  for (let num of nums) {
    array[num + max] = 1;
  }
  let pointer = 0,
    count = 0,
    ans = 1;
  // traverse array with pointer and count
  while (pointer < array.length) {
    if (array[pointer] == 1) {
      count++;
    } else {
      ans = Math.max(ans, count);
      count = 0;
    }
    pointer++;
  }
  return Math.max(ans, count);
};

File: hash_table/0128_longest_consecutive_sequence.py
# Given an unsorted array of integers nums, return the length of the longest 
# consecutive elements sequence.

# You must write an algorithm that runs in O(n) time.

# https://leetcode.com/problems/longest-consecutive-sequence/

# sort
# time complexity O(nlogn)
# space complexity O(n)
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        if not nums: 
            return 0
        sorted_num = sorted(list(set(nums)))
        ans = cnt = 1
        for i in range(1, len(sorted_num)):
            if sorted_num[i] - sorted_num[i - 1] == 1:
                cnt += 1
                ans = max(ans, cnt)
            else: 
                cnt = 1
        return ans
    
# hash table -- MLE
# time complexity O(n)
# space complexity O(n)
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        if len(nums) <= 1:
            return len(nums)
        mi, ma = min(nums), max(nums)
        track = [' ' for _ in range(ma - mi + 1)]
        for num in nums:
            track[num - mi] = '1'
        res = ''.join(track).split()
        max_cnt = 0
        for ele in res:
            max_cnt = max(len(ele), max_cnt)
        return max_cnt

# hash table
# time complexity O(n)
# space complexity O(n)
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        longest_streak = 0
        num_set = set(nums)
        
        for num in num_set:
            if num - 1 not in num_set:  # num is the start of consecutive sequence
                current_num = num
                current_streak = 1
                while current_num + 1 in num_set:
                    current_num += 1
                    current_streak += 1
                longest_streak = max(longest_streak, current_streak)
        
        return longest_streak

